### PR TITLE
Introduce hover actions

### PR DESCRIPTION
**Pull request by vsrs**
The idea discussed on [Zulip](https://rust-lang.zulipchat.com/#narrow/stream/185405-t-compiler.2Fwg-rls-2.2E0/topic/Hover.20actions)

Screencasts:
1) Runnable actions
hover_actions_run

2) Go to Type Definition:
hover_actions_goto

3) And to be able to completely disable code lenses I've added a `Go to Implementations` action:
hover_actions_impl

P.S. I don't forget about ```struct Markdown(String)```. It's just that such a change affects much more code than I expected, and I think it's better to do it in a separate PR.

### CODE REVIEW

**Comment by aloucks** [comment]
Does this add new functionality or just move the code lens into the tooltip?

**Comment by vsrs** [comment]
For inlay hints yes, it does. It makes possible to choose a type to go to.

**Comment by aloucks** [comment]
> For inlay hints yes, it does. It makes possible to choose a type to go to.

These look really nice and definitely useful!

However, I think I prefer the way the code lens was displayed before (outside of tooltip) for running tests and finding implementations.

**Comment by vsrs** [comment]
HoverActions is not a replacement for CodeLens. It's an alternative. You can even use both :) It's all configurable. By default in vscode all lenses and hover actions are enabled.

**Comment by kjeremy** [comment]
I haven't had time to really look at this but I think you should open an issue upstream in the vscode repo so they can consider a story for supporting this capability for all extensions.

Alternatively would it be more natural to return `Command[]` in a hover response from the LSP server? If so maybe an issue should be opened [here](https://github.com/microsoft/language-server-protocol) to allow this capability.

**Comment by vsrs** [comment]
> Alternatively would it be more natural to return `Command[]` in a hover response from the LSP server?
Good idea! Thanks. I'll open an issue in the LSP repo.

**Comment by kjeremy** [comment]
> 
> 
> > Alternatively would it be more natural to return `Command[]` in a hover response from the LSP server?
> 
> Good idea! Thanks. I'll open an issue in the LSP repo.

It may be worth getting this into the form of our other "custom lsp extensions" (see [code action groups](https://github.com/rust-analyzer/rust-analyzer/blob/master/docs/dev/lsp-extensions.md#codeaction-groups)) as an example to prove that it would work upstream. Then you can open an issue in the LSP repo and point to our implementation.

**Comment by vsrs** [comment]
>Then you can open an issue in the LSP repo and point to our implementation.

I'm not sure it is possible to fully implement it without digging into vscode.

Let me explain. If we add hover actions in the protocol I think thay should be a first-class vscode citizens and appear exactly the same way as vscode actions:
hover_actions

1 - Where it is possible to insert command links using existing API.
2 - Where they should be to my mind.

BTW, having command links in markdown has one benefit: vscode supports command links from 2017 (https://github.com/microsoft/vscode/pull/29082) and there is good chance that other editors support them too (I did not check).

Anyway. I'll try to reimplement hover actions using Command[] in another branch to prove the idea. And then we can select between the two.

Please do not merge this PR until then.

**Comment by kjeremy** [comment]
> BTW, having command links in markdown has one benefit: vscode supports command links from 2017 ([microsoft/vscode#29082](https://github.com/microsoft/vscode/pull/29082)) and there is a good chance that other editors support them too (I didn't check).

Ah cool! I wasn't aware of that.

**Comment by aloucks** [comment]
> > BTW, having command links in markdown has one benefit: vscode supports command links from 2017 ([microsoft/vscode#29082](https://github.com/microsoft/vscode/pull/29082)) and there is a good chance that other editors support them too (I didn't check).
> 
> Ah cool! I wasn't aware of that.

I wonder if this could be use to transform [intra-doc](https://github.com/rust-lang/rfcs/blob/master/text/1946-intra-rustdoc-links.md) links into `go-to-def` links?

**Comment by vsrs** [comment]
> I wonder if this could be use to transform intra-doc links into go-to-def links?

Yes, it could.

This PR adds `gotoLocation` command: [client side](https://github.com/vsrs/rust-analyzer/blob/7f990539faa4ad9534aa0c5120ee6f6a4fa80eeb/editors/code/src/commands.ts#L358), [server side](https://github.com/vsrs/rust-analyzer/blob/7f990539faa4ad9534aa0c5120ee6f6a4fa80eeb/crates/rust-analyzer/src/main_loop/handlers.rs#L1039)

I see no problems to find all intra-doc links in the server hover handler and replace them with command links.

For example:
`The offers several ways to fooify [Bars](bars::Bar).` ->
`The offers several ways to fooify [Bars](command:rust-analyzer.gotoLocation?...).`

should be fairly easy.

**Comment by vsrs** [comment]
> And then we can choose between two.

The LSP extension implementation is ready. I think this way is far better and there is no point in comparing it with the old one.

File: crates/ra_hir/src/code_model.rs [hunk 1]
```diff
@@ -1333,6 +1333,11 @@ impl Type {
         Some(adt.into())
     }
 
+    pub fn substs(&self) -> Option<Substs> {
+        let (_adt, subst) = self.ty.value.as_adt()?;
+        Some(subst.clone())
+    }
```
Comment (matklad) [inline crates/ra_hir/src/code_model.rs:1339]: cc @flodiebold , not sure if want to expose `Subst` from `hir::Type` or not...

My gut feeling is that we want to keep this private, but I don't have a good understanding here...

File: crates/ra_ide/src/hover.rs [hunk 2]
```diff
@@ -18,14 +18,64 @@ use ra_syntax::{
...
+pub struct HoverConfig {
+    pub run: bool,
+    pub debug: bool,
+    pub goto_type_def: bool,
+    pub implementations: bool,
+}
+
+impl Default for HoverConfig {
+    fn default() -> Self {
+        // A client should explicitly specify if it supports hover actions!
```
Comment (matklad) [inline crates/ra_ide/src/hover.rs:36]: I think in `ra_ide` we generally enable everything by deafault, and than `rust-analyzer` explicitelly disables stuff it doesn't like.

File: crates/ra_ide/src/hover.rs [hunk 3]
```diff
@@ -18,14 +18,64 @@ use ra_syntax::{
...
+
+    pub fn none(&self) -> bool {
+        !self.any()
+    }
+}
+
+#[derive(Debug, Clone, Eq, PartialEq)]
+pub struct HoverGotoTypeData {
+    pub hint: String,
+    pub link: NavigationTarget,
```
Comment (matklad) [inline crates/ra_ide/src/hover.rs:58]:
```suggestion
    pub nav: NavigationTarget,
```

File: crates/ra_ide/src/hover.rs [hunk 4]
```diff
@@ -18,14 +18,64 @@ use ra_syntax::{
...
+    }
+
+    pub fn none(&self) -> bool {
+        !self.any()
+    }
+}
+
+#[derive(Debug, Clone, Eq, PartialEq)]
+pub struct HoverGotoTypeData {
+    pub hint: String,
```
Comment (matklad) [inline crates/ra_ide/src/hover.rs:57]: the meaning of `hint` is not immediately obvious to me, so a comment (or maybe a rename) is wanted.

File: crates/ra_ide/src/hover.rs [hunk 5]
```diff
@@ -53,10 +103,20 @@ impl HoverResult {
         &self.results
     }
 
+    pub fn actions(&self) -> &[HoverAction] {
+        &self.actions
+    }
+
+    pub fn append_action(&mut self, action: HoverAction) {
```
Comment (matklad) [inline crates/ra_ide/src/hover.rs:110]:
```suggestion
    pub fn push_action(&mut self, action: HoverAction) {
```

File: crates/ra_ide/src/hover.rs [hunk 6]
```diff
@@ -53,10 +103,20 @@ impl HoverResult {
...
+        self.actions.push(action);
+    }
+
     /// Returns the results converted into markup
     /// for displaying in a UI
+    ///
+    /// Does not process actions!
     pub fn to_markup(&self) -> String {
-        self.results.join("\n\n---\n")
+        self.results.join("\n\n___\n")
```
Comment (matklad) [inline crates/ra_ide/src/hover.rs:119]: Weren't we changing this back & forth a couple of PRs ago?

File: crates/ra_ide/src/hover.rs [hunk 7]
```diff
@@ -215,6 +288,77 @@ fn hover_text_from_name_kind(db: &RootDatabase, def: Definition) -> Option<Strin
...
+        Definition::ModuleDef(it) => match it {
+            ModuleDef::Adt(Adt::Struct(it)) => Some(to_impl_action(it.to_nav(db))),
+            ModuleDef::Adt(Adt::Union(it)) => Some(to_impl_action(it.to_nav(db))),
+            ModuleDef::Adt(Adt::Enum(it)) => Some(to_impl_action(it.to_nav(db))),
+            ModuleDef::Trait(it) => Some(to_impl_action(it.to_nav(db))),
+            _ => None,
+        },
+        _ => None,
+    }
+}
```
Comment (matklad) [inline crates/ra_ide/src/hover.rs:338]: I feel this is the most complex action here. Would you mind moving it to a separate PR, so that we first can review the general LSP interface changes, and then tackle the specifics of type infernce and such? As per our would-be-introduce contribution guidelines :-) 

https://github.com/rust-analyzer/rust-analyzer/pull/4703/files#diff-83b898006efcb044dce73a8a5cc94ed6R156-R161

File: crates/ra_ide/src/hover.rs (hunk 2 above)
Comment (vsrs) [reply crates/ra_ide/src/hover.rs:36]: ok. Besides, a I've added a hoverActions client capability, so this initial defaults outdated.

File: crates/ra_ide/src/hover.rs (hunk 4 above)
Comment (vsrs) [reply crates/ra_ide/src/hover.rs:57]: It's a tooltip for the command link, when represented in the UI. I'll rename.

File: crates/ra_ide/src/hover.rs (hunk 6 above)
Comment (vsrs) [reply crates/ra_ide/src/hover.rs:119]: As far as I can see there are some places where '---' were replaced by '___' but not this one.

File: crates/ra_ide/src/hover.rs (hunk 7 above)
Comment (vsrs) [reply crates/ra_ide/src/hover.rs:338]: Sure, I'll close this PR and split it to a several ones.

File: crates/ra_hir/src/code_model.rs (hunk 1 above)
Comment (vsrs) [reply crates/ra_hir/src/code_model.rs:1339]: > My gut feeling is that we want to keep this private, but I don't have a good understanding here...

Then what is the right way go get all type parameters?

File: crates/ra_ide/src/hover.rs (hunk 4 above)
Comment (vsrs) [reply crates/ra_ide/src/hover.rs:57]:
```suggestion
    pub tooltip: String,
```

File: crates/ra_hir/src/code_model.rs (hunk 1 above)
Comment (vsrs) [reply crates/ra_hir/src/code_model.rs:1339]: > For the current UI, I could imagine just producing a flat list of all the ADTs mentioned in the type (using walk that would be easy)

@flodiebold are you talking about https://github.com/rust-analyzer/rust-analyzer/blob/0035dafbfa563921e2cfe41f5592fc37bff92294/crates/ra_hir_ty/src/lib.rs#L844-L845 or I got something wrong? Looks like TypeWalk is implemented for `Ty`, `Substs`, etc, not for `Type`. 

Could you please give me some pointers to the existing code that does something similar? I tried to understand how inlay hints (via HirDisplay\HirFormatter?) work, but at the moment it is too hard for me :( Using substs was my bets bet.

Comment (flodiebold) [reply crates/ra_hir/src/code_model.rs:1339]: Yes, that's what I meant. You're right that it isn't implemented for `Type`, which makes sense because like `Subst`, it exposes internals of `hir_ty`, which we don't want to do in the `hir` facade.

What I was imagining is adding a method on `Type` for this; either a very specific one that uses `walk` internally and returns a `Vec<Adt>`, or maybe even a full `walk` function (but one that passes `Type`s instead of `Ty`s).

Comment (vsrs) [reply crates/ra_hir/src/code_model.rs:1339]: I see. I'll try to get simple flat list for the start.
Thanks.
